### ISSUE TITLE
SIGCHLD causes "Interrupted system call" errors on Python 2.4, in a fresh SVN checkout

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4331]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| TimAllen reported|
|-|-|
|Trac ID|trac#4331|
|Type|defect|
|Created|2010-03-01 03:34:30Z|

Apparently this is a regression from [#10460](https://github.com/twisted/twisted/issues/10460).

From a fresh `git svn` checkout of r28580 on my Python 2.4-based system, when I run `bin/trial twisted/` I get the following errors:
```
===============================================================================
[FAIL]: twisted.internet.test.test_process.PTYProcessTestsBuilder_SelectReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 219, in test_systemCallUninterruptedByChildExit
    self.assertEqual(result, ["Foo\n"])
twisted.trial.unittest.FailTest: not equal:
a = []
b = ['Foo\n']

===============================================================================
[FAIL]: twisted.internet.test.test_process.ProcessTestsBuilder_PollReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 219, in test_systemCallUninterruptedByChildExit
    self.assertEqual(result, ["Foo\n"])
twisted.trial.unittest.FailTest: not equal:
a = []
b = ['Foo\n']

===============================================================================
[FAIL]: twisted.internet.test.test_process.ProcessTestsBuilder_SelectReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 219, in test_systemCallUninterruptedByChildExit
    self.assertEqual(result, ["Foo\n"])
twisted.trial.unittest.FailTest: not equal:
a = []
b = ['Foo\n']

===============================================================================
[ERROR]: twisted.internet.test.test_process.PTYProcessTestsBuilder_PollReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/base.py", line 409, in _continueFiring
    callable(*args, **kwargs)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 213, in f
    result.append(f2.read())
exceptions.IOError: [Errno 4] Interrupted system call
===============================================================================
[ERROR]: twisted.internet.test.test_process.PTYProcessTestsBuilder_SelectReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/base.py", line 409, in _continueFiring
    callable(*args, **kwargs)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 213, in f
    result.append(f2.read())
exceptions.IOError: [Errno 4] Interrupted system call
===============================================================================
[ERROR]: twisted.internet.test.test_process.ProcessTestsBuilder_PollReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/base.py", line 409, in _continueFiring
    callable(*args, **kwargs)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 213, in f
    result.append(f2.read())
exceptions.IOError: [Errno 4] Interrupted system call
===============================================================================
[ERROR]: twisted.internet.test.test_process.ProcessTestsBuilder_SelectReactor.test_systemCallUninterruptedByChildExit

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/base.py", line 409, in _continueFiring
    callable(*args, **kwargs)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/internet/test/test_process.py", line 213, in f
    result.append(f2.read())
exceptions.IOError: [Errno 4] Interrupted system call
-------------------------------------------------------------------------------
```

Ticket [#10460](https://github.com/twisted/twisted/issues/10460) comment 42 says "If neither of these is available, revert to the old behavior." It seems the revert-to-old-behaviour part of the patch isn't really working.

(as a side note, if there's a way to run Twisted from an SVN checkout with all the C extensions built, I'd be interested to hear it)

Attachments:

* [4331.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9a2/9a2a493fdc6132a2de59f64b42b3edebb0540e5d/bd47104056f753b1e47b645d02bd8da7487c14d3.patch) (1033 bytes) - added by michaelnt on 2011-01-31 02:38:12Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4331 4331
type__defect defect
reporter__TimAllen TimAllen
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1267414470000000 1267414470000000
changetime__1297668661000000 1297668661000000
version__None None
owner__ 
cc__spiv cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

For the time being, you will want to do `python setup.py build_ext --inplace`.

Are you sure this is a regression?  As I understand it, in principle, those errors were always possible.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|TimAllen commented|
|-|-|

Looking more closely, it seems the failing tests in question didn't exist before [#10460](https://github.com/twisted/twisted/issues/10460) was merged, so I guess it's not technically a regression. Running the `build_ext` command you suggest makes the tests pass, as expected.

Still, it would be nice if the tests were SKIPPED or TODO rather than FAIL and ERROR in an environment where we know they won't work.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to TimAllen|
|-|-|

I think a first cut at fixing this would just be to skip these tests if Python is older than 2.6 and the native module isn't available for some reason.  Care to take a whack at it?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Skipping sounds good.  Shouldn't be too tricky to implement.  A Python version check is easy, of course.  For the rest, just try the import I guess and skip if you get an `ImportError`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

[#4466](https://github.com/twisted/twisted/issues/4466) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|michaelnt commented|
|-|-|

Is this issue to do with not building the c extensions on a fresh checkout or is it to do with Python 2.4. If it is to do with not building the extensions is it worth fixing?

If I try reproduce this error and run for instance t.i.t.test_process.PTYProcessTestsBuilder_Glib2Reactor the tests just hang.

It's not very clear from the previous comments what should be done to resolve this issue.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|Screwtape removed owner|
|-|-|

I believe it is both - Python 2.6 has native support for this functionality; older Pythons don't. Pre-packaged Twisted has a C extension to support this functionality; SVN checkouts don't. Therefore people in the intersection of "people using Python 2.4" and "people using Twisted from SVN" (basically: people trying to get Twisted working smoothly on RHEL) have a problem.

Glyph suggested the fix "skip tests if the native functionality is missing and the C extension is missing", and no better suggestions have been made yet.

(for the record, TimAllen was my old username, and I am not currently working on this, so removing ownership)

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|michaelnt commented|
|-|-|

I've attached a patch that skips the tests.

Would it be better to add an attribute to the _signals module to indicate more clearly if this feature is supported?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_12"></a>|@jerub commented|
|-|-|

I have not been able to replicate the hang on linux, python2.4, select reactor, without the patch, but I have verified that the patch does skip tests correctly when the c library is not available.

Thank-you, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_13"></a>|@jerub set status to `closed`|
|-|-|

(In [30753]) SIGCHLD causes "Interrupted system call" errors on Python 2.4, in a fresh SVN checkout

Author: michaelnt
Reviewer: jerub
Fixes: [#4331](https://github.com/twisted/twisted/issues/4331)

When C-extensions are not complied for twisted, on python2.4, skip a test in
twisted.internet.test.test_process that may hang due to a SIGCHLD related
problem. Running 'python setup.py build_ext --inplace' will compile the
extension and cause the test to both run and pass.
